rfd/0059-search-based-access-requests.md

```diff
@@ -0,0 +1,276 @@
+
+Alice needs to locate the "db-1" resource and request access to it.
+
+Alice goes to the “Request access” screen and selects: “select access to resources”
+option that is available alongside the “select access to roles” option.
+
+Alice uses standard (fuzzy search) or advanced search (labels) to enter the
+resource name.
+
+She finds a single resource (e.g. "db-1") or a group of resources by that name and clicks
+"request access".
+
+An access request is created. It is sent to the "db-admins" team members via slack,
+email or other configured channel based on the labels on the requested resources
+(we will add label-based routing to access request plugins).
+
+There are two thresholds set in access requests for this group of resources.
+When both Ivan and Mary approve the request, Alice is granted access for one
+hour.
+
+### Query Language and Filter UI
+
+The query language and filter UI is defined by the
+[Pagination and Search RFD](./0055-webui-ss-paginate-filter.md).
+
+### CLI Mock-Up
+
+```bash
+$ tsh request search --kind db --labels 'env=prod'
+Found 2 items:
+
+name kind     id
+db-1 database 388aff7f-459f-4a43-804a-3729854976ab
+db-2 database 3be2fdad-7c79-4cfa-924e-ec1ea7225320
+
+Create access request by:
+> tsh request create --id 388aff7f-459f-4a43-804a-3729854976ab --id 3be2fdad-7c79-4cfa-924e-ec1ea7225320
+```
+
+Users can search by kind, labels, and keywords. The `tsh request search` command
+will output a `tsh request create` command which can be used to request access.
+
+The CLI UX may be improved by allowing users to perform multiple searches,
+"stage" which resources they want to request access to and store this state
+locally, and finally execute the full request without needing each UUID on the
+command line. But this may be too complex and confusing. With this MVP version
+users can copy-paste UUIDs from multiple searches, the meaning is clear, and it
+is fully customizable and scriptable.
+
+### On-Demand SSH
+
+Many times users would not want to search and request access in two steps.
+Teleport will have a new alias flag modifier that will try accessing a node, in
+case if access denied, it will request access for a node all in one step:
+
+```bash
+# -P "Please" activate try and request mode
+$ tsh ssh -P root@db-node
+# by default, if tsh will succeed to SSH, it will just continue with a
+# Session, otherwise it will find a node and create a search based access request
+You do not have access to the system by default, created access request.
+
+Please wait...
+
+Access request has been approved. Happy hacking!
+$
+```
+
+Some customers would want to have `-P` flag as a default. New alias feature in
+profile, will allow those users to set `tsh ssh` as an alias for `tsh ssh -P`.
+
+For OpenSSH use-cases, some users would like to load the new certificates in the
+agent, `tsh login -A root@db-node` will work like `tsh ssh -P` command except
+instead of login, it will load the keys in the agent.
```

Comment:
This would obviously be great for some users, but this to me is a clear future/v2 feature.